File: session/03_visitor/haskell/subsumed_visitor.hs
data Shape = Cylinder {
  baseRadius :: Double,
  height :: Double
} | Sphere {
  radius :: Double
} deriving (Show)

volume :: Shape -> Double
volume (Cylinder baseRadius height) = pi * baseRadius ^^ 2 * height
volume (Sphere radius) = 4.0 / 3.0 * pi * radius ^^ 3

surfaceArea :: Shape -> Double
surfaceArea (Cylinder baseRadius height) = 2 * baseArea + baseCircumference * height
  where
    baseArea = pi * baseRadius ^^ 2
    baseCircumference = 2 * pi * baseRadius

surfaceArea (Sphere radius) = 4.0 * pi * radius ^^ 2

main :: IO ()
main = do
  let cylinder = Cylinder 10 10
  let sphere = Sphere 10
  print $ "Volume of Cylinder " ++ show (volume cylinder)
  print $ "Volume of Sphere " ++ show (volume sphere)
  print $ "Surface Area of Cylinder " ++ show (surfaceArea cylinder)
  print $ "Surface Area of Sphere " ++ show (surfaceArea sphere)

File: session/03_visitor/erlang/subsumed_visitor.erl
#!/usr/bin/env escript

-record (cylinder, {baseRadius = 0, height = 0}).
-record (sphere, {radius = 0}).

volume(#cylinder {baseRadius = BaseRadius, height = Height}) ->
	math:pi() * math:pow(BaseRadius, 2) * Height;
	
volume(#sphere {radius = Radius}) ->
	4.0 / 3.0 * math:pi() * math:pow(Radius, 3);
	
volume(_) -> 
    not_ok.

surfaceArea(#cylinder {baseRadius = BaseRadius, height = Height}) ->
	BaseArea = math:pi() * math:pow(BaseRadius, 2),
	BaseCircumference = 2 * math:pi() * BaseRadius,
	2 * BaseArea + BaseCircumference * Height;
	
surfaceArea(#sphere {radius = Radius}) ->
	4.0 * math:pi() * math:pow(Radius, 2);
	
surfaceArea(_) -> 
    not_ok.
	
main([]) ->
  Cylinder = #cylinder {baseRadius = 10, height = 10},
  Sphere = #sphere {radius = 10},
  io:format("Volume of Cylinder ~p~n", [volume(Cylinder)]),
  io:format("Volume of Sphere ~p~n", [volume(Sphere)]),
  io:format("Volume of 2 ~p~n", [volume(2)]),
  io:format("Surface Area of Cylinder ~p~n", [surfaceArea(Cylinder)]),
  io:format("Surface Area of Sphere ~p~n", [surfaceArea(Sphere)]),
  io:format("Surface Area of 2 ~p~n", [surfaceArea(2)]).